src/fetchers/tarball-fetcher.js

```diff
@@ -130,48 +131,66 @@ export default class TarballFetcher extends BaseFetcher {
-          // should we save this to the offline cache?
-          const tarballMirrorPath = this.getTarballMirrorPath();
-          const tarballCachePath = this.getTarballCachePath();
-
-          const {validateStream, extractorStream} = this.createExtractor(resolve, reject);
-
-          const handleRequestError = res => {
-            if (res.statusCode >= 400) {
-              const statusDescription = http.STATUS_CODES[res.statusCode];
-              reject(new Error(reporter.lang('requestFailed', `${res.statusCode} ${statusDescription}`)));
-            }
-          };
-
-          req.on('response', handleRequestError);
-          req.pipe(validateStream);
-
-          if (tarballMirrorPath) {
-            validateStream.pipe(fs.createWriteStream(tarballMirrorPath)).on('error', reject);
-          }
-
-          if (tarballCachePath) {
-            validateStream.pipe(fs.createWriteStream(tarballCachePath)).on('error', reject);
-          }
-
-          validateStream.pipe(extractorStream).on('error', reject);
-        },
-      },
-      this.packageName,
-    );
+    let retriesRemaining = 2;
+    do {
+      try {
+        return await registry.request(
+          this.reference,
+          {
+            headers: {
+              'Accept-Encoding': 'gzip',
+              Accept: 'application/octet-stream',
+            },
+            buffer: true,
+            process: (req, resolve, reject) => {
+              // should we save this to the offline cache?
+              const {reporter} = this.config;
+              const tarballMirrorPath = this.getTarballMirrorPath();
+              const tarballCachePath = this.getTarballCachePath();
+
+              const {validateStream, extractorStream} = this.createExtractor(resolve, reject);
+
+              req.on('response', res => {
+                if (res.statusCode >= 400) {
+                  const statusDescription = http.STATUS_CODES[res.statusCode];
+                  reject(
+                    new ResponseError(
+                      reporter.lang('requestFailed', `${res.statusCode} ${statusDescription}`),
+                      res.statusCode,
+                    ),
+                  );
+                }
+              });
+              req.pipe(validateStream);
+
+              if (tarballMirrorPath) {
+                validateStream.pipe(fs.createWriteStream(tarballMirrorPath)).on('error', reject);
+              }
+
+              if (tarballCachePath) {
+                validateStream.pipe(fs.createWriteStream(tarballCachePath)).on('error', reject);
+              }
+
+              validateStream.pipe(extractorStream).on('error', reject);
+            },
+          },
+          this.packageName,
+        );
+      } catch (err) {
+        if (err instanceof ResponseError && err.responseCode >= 500 && retriesRemaining > 1) {
+          retriesRemaining--;
+          this.reporter.warn(this.reporter.lang('retryOnInternalServerError'));
+          await sleep(3000);
```

Comment:
Same as above, should we pull this out to `RETRY_DELAY`? We actually have this same delay in a few places now.